File: Days 31 - 40/Day 35/SortRGBArray.cpp
#include <array>
#include <iostream>
#include <utility>

template <typename T, unsigned int Size>
int PullElementsToFront(std::array<T, Size>& array, unsigned int startIndex, unsigned int endIndex, const T& element) noexcept
{
	int currentIndex = -1;

	while (startIndex < endIndex)
	{
		if (array[startIndex] == element)
		{
			currentIndex = startIndex;
			startIndex++;
		}
		else if (array[endIndex] != element)
		{
			endIndex--;
		}
		else
		{
			currentIndex = startIndex;

			std::swap(array[startIndex], array[endIndex]);
		}
	}

	return currentIndex;
}

template <unsigned int Size>
std::array<char, Size> SortRGBArray(const std::array<char, Size>& array)
{
	static constexpr unsigned int TOTAL_ELEMENTS = 3;
	static constexpr char ELEMENTS[TOTAL_ELEMENTS] { 'R', 'G', 'B' };

	std::array<char, Size> sortedArray(array);
	int partition = -1;

	for (unsigned int i = 0; i < TOTAL_ELEMENTS - 1; i++)
	{
		partition = PullElementsToFront(sortedArray, partition + 1, sortedArray.size() - 1, ELEMENTS[i]);
	}

	return sortedArray;
}

template <typename T, unsigned int Size>
void PrintArray(const std::array<T, Size>& array)
{
	for (const auto& value : array)
	{
		std::cout << value << " ";
	}

	std::cout << "\n";
}

int main(int argc, const char* argv[])
{
	std::array<char, 7> array { 'G', 'B', 'R', 'R', 'B', 'R', 'G' };

	std::cout << "Original array: ";
	PrintArray(array);

	std::cout << "  Sorted array: ";
	PrintArray(SortRGBArray(array));

	std::cin.get();

	return 0;
}

File: Days 31 - 40/Day 35/SortRGBArray.cs
using System;

namespace DailyCodingProblem
{
	internal class Day35
	{
		private static int Main(string[] args)
		{
			char[] array = new char[] { 'G', 'B', 'R', 'R', 'B', 'R', 'G' };
			Console.Write("Original array: ");
			PrintArray(array);

			Console.Write("  Sorted array: ");
			PrintArray(SortRGBArray(array));

			Console.ReadLine();

			return 0;
		}

		private static char[] SortRGBArray(char[] array)
		{
			const int TotalElements = 3;
			char[] Elements = new char[TotalElements] {'R', 'G', 'B' };

			char[] sortedArray = new char[array.Length];
			Array.Copy(array, sortedArray, array.Length);

			int partition = -1;

			for (int i = 0; i < TotalElements; i++)
			{
				partition = PullElementsToFront(sortedArray, partition + 1, sortedArray.Length - 1, Elements[i]);
			}

			return sortedArray;
		}

		private static int PullElementsToFront(char[] array, int startIndex, int endIndex, char element)
		{
			int currentIndex = -1;

			while (startIndex < endIndex)
			{
				if (array[startIndex] == element)
				{
					currentIndex = startIndex;
					startIndex++;
				}
				else if (array[endIndex] != element)
				{
					endIndex--;
				}
				else
				{
					currentIndex = startIndex;

					char temp = array[startIndex];
					array[startIndex] = array[endIndex];
					array[endIndex] = temp;
				}
			}

			return currentIndex;
		}

		private static void PrintArray<T>(T[] array)
		{
			foreach (T value in array)
			{
				Console.Write($"{value} ");
			}

			Console.WriteLine();
		}
	}
}